examples/cuda/riscv.mk

```diff
@@ -49,41 +61,140 @@ RISCV_GNU_PATH   := $(RISCV_TOOLS_PATH)/riscv-install
+# infer the correct scheduling weights for important HB features like
+# bsg_attr_remote because -mcpu and other flags are not passed from
+# clang (the front end). See:
+# https://github.com/bespoke-silicon-group/bsg_manycore/blob/3a05e000aef1fba896f756e9836c9e002e5a06f7/software/mk/Makefile.builddefs#L208
+
+# As a result clang/LLVM compilation is split into
+# multiple steps:
+#   1. Clang (.c/.cpp -> .ll (LLVM IR))
+#   2. llc (.ll (LLVM IR) -> .s)
+#   3. gcc/clang (.s -> .rvo)
+
+# This is problematic if we want to selectively compile objects with
+# clang/LLVM because it adds intermediate targets that GCC does not
+# have. AFAIK there is no easy way to express two different
+# compilation paths in a single run of make. If clang correctly passed
+# the arguments to lcc and HB RV32 objects could be optimally comiled
+# with a single clang command, this issue would be resolved and we
+# could use a flow like the x86 example above.
+
+# The current approach in bsg_manycore removes the ability to do
+# per-object compiler selection. CLANG=1 is set globally, which
+# compiles ALL application objects with LLVM, or none of them.
+
+# Hence, this solution. The approach below seems to be the only way to
+# provide per-object selection of GCC/LLVM with our current flow.
+#
+# Instead of having a single command, we wrap clang/LLVM into a
+# makefile function to provide the abstraction of a single
+# command. These functions are RISCV_GCC/RISCV_GXX or
+# RISCV_CLANG/RISCV_CLANGXX for C and C++ respectively.
+#
+# Providing RISCV_GCC/RISCV_GXX and RISCV_CLANG/RISCV_CLANGXX as
+# callable functions provides the abstraction of being able to choose
+# the compiler on a per-object basis (e.g. kernel.rvo:RISCV_CC=RISC_GCC) or globally.
+
+# For an example of how to use this file see
+# test_credit_csr_tile/Makefile.
+
+RISCV_GCC = $(_RISCV_GCC) $(RISCV_CFLAGS) $(RISCV_DEFINES) $(RISCV_INCLUDES) -c $< -o $@ |& tee $*.rvo.log
+RISCV_GXX = $(_RISCV_GXX) $(RISCV_CXXFLAGS) $(RISCV_DEFINES) $(RISCV_INCLUDES) -c $< -o $@ |& tee $*.rvo.log
+
+# The clang version of the functions hide the fact that LLC (the
+# optimizer) is being called between steps.
+define RISCV_CLANGXX
+$(_RISCV_CLANGXX) $(RISCV_CXXFLAGS) $(RISCV_DEFINES) $(RISCV_INCLUDES) -c $< -o $@.ll -S -emit-llvm && \
+$(RISCV_LLVM_LLC) $(RISCV_LLVM_LLC_FLAGS) $@.ll -o $@.S && \
+$(_RISCV_CLANG) $(RISCV_CFLAGS) $(RISCV_DEFINES) $(RISCV_INCLUDES) -c $@.S -o $@ |& tee $*.rvo.log
+endef 
+
+define RISCV_CLANG
+$(_RISCV_CLANG) $(RISCV_CFLAGS) $(RISCV_DEFINES) $(RISCV_INCLUDES) -c $< -o $@.ll -S -emit-llvm |& tee $*.rvo.ll.log && \
+$(RISCV_LLVM_LLC) $(RISCV_LLVM_LLC_FLAGS) $@.ll -o $@.S |& tee $*.rvo.S.log && \
+$(_RISCV_CLANG) $(RISCV_CFLAGS) $(RISCV_DEFINES) $(RISCV_INCLUDES) -c $@.S -o $@ |& tee $*.rvo.log
+endef
+
+# Set the default RISC-V Compilers.
 RISCV_CXX ?= $(RISCV_GXX)
 RISCV_CC  ?= $(RISCV_GCC)
 
+# To chose between GCC and CLANG set RISCV_CC to RISCV_GCC or RISCV_CLANG, respectively.
```

Comment:
I don't think the use of ", respectively" is needed here (or correct?)

Reply:
I think it is